Plutarch/DataRepr/Internal.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans -Wno-redundant-constraints #-}
+
+module Plutarch.DataRepr.Internal (
+  PDataSum,
+  punDataRepr,
+  pindexDataRepr,
+  pmatchDataRepr,
+  DataReprHandlers (..),
+  PDataRecord,
+  PLabeledType (..),
+  type PUnLabel,
+  type PLabel,
+  pdhead,
+  pdtail,
+  PIsDataRepr (..),
+  PIsDataReprInstances (..),
+  pindexDataRecord,
+  pdropDataRecord,
+  DerivePConstantViaData (..),
+) where
+
+import Data.List (groupBy, maximumBy, sortOn)
+import Data.Proxy (Proxy)
+import GHC.TypeLits (KnownNat, Symbol, natVal)
+import Numeric.Natural (Natural)
+import Plutarch (Dig, PMatch, TermCont, hashOpenTerm, punsafeBuiltin, punsafeCoerce, runTermCont)
+import Plutarch.Bool (pif, (#==))
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PData,
+  PIsData,
+  pasConstr,
+  pdata,
+  pfromData,
+  pfstBuiltin,
+  psndBuiltin,
+ )
+import Plutarch.DataRepr.Internal.HList (type Drop, type IndexList)
+import Plutarch.Integer (PInteger)
+import Plutarch.Lift (PConstant, PConstantRepr, PConstanted, PLift, pconstantFromRepr, pconstantToRepr)
+import Plutarch.List (pdrop, punsafeIndex)
+import Plutarch.Prelude
+import qualified Plutus.V1.Ledger.Api as Ledger
+import qualified PlutusCore as PLC
+
+data PDataRecord (as :: [PLabeledType]) (s :: S)
+
+data PLabeledType = Symbol := PType
+
+pdhead :: Term s (PDataRecord ((l ':= a) : as) :--> PAsData a)
+pdhead = phoistAcyclic $ pforce $ punsafeBuiltin PLC.HeadList
+
+pdtail :: Term s (PDataRecord (a ': as) :--> PDataRecord as)
+pdtail = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
+
+type family PUnLabel (as :: [PLabeledType]) :: [PType] where
+  PUnLabel '[] = '[]
+  PUnLabel ((l ':= t) ': as) = t ': (PUnLabel as)
+
+type family PLabel (as :: [PLabeledType]) :: [Symbol] where
+  PLabel '[] = '[]
+  PLabel ((l ':= t) ': as) = l ': (PLabel as)
+
+type PDataSum :: [[PLabeledType]] -> PType
+data PDataSum (defs :: [[PLabeledType]]) (s :: S)
+
+pasData :: Term s (PDataSum _) -> Term s PData
+pasData = punsafeCoerce
+
+punDataRepr :: Term s (PDataSum '[def] :--> PDataRecord def)
+punDataRepr = phoistAcyclic $
+  plam $ \t ->
+    plet (pasConstr #$ pasData t) $ \d ->
+      (punsafeCoerce $ psndBuiltin # d :: Term _ (PDataRecord def))
+
+pindexDataRepr :: (KnownNat n) => Proxy n -> Term s (PDataSum (def : defs) :--> PDataRecord (IndexList n (def : defs)))
```

Comment:
Dead code?

Reply:
I think it should be exported, incase there's a use for it, though most of the time pmatch would work instead

Reply:
It has the same function as `pindexDataRecord` though?